### PR TITLE
for shared-memory handles, use an atomic counter, instead of potentially colliding random numbers

### PR DESCRIPTION
**Pull request by mbc**
These handles, used for shared-memory tensors, can collide.

E.g. see https://github.com/pytorch/pytorch/issues/60626#issuecomment-869919018

### CODE REVIEW

File: torch/csrc/generic/StorageSharing.cpp [hunk 1]
```diff
@@ -42,15 +42,15 @@ static PyObject * THPStorage_(sharedIncref)(PyObject *_self, PyObject *noargs)
 #ifndef THC_GENERIC_FILE
 // TODO: move this somewhere - we only need one version
 static std::string THPStorage_(__newHandle)() {
-  static std::random_device rd;
+  static std::atomic<unsigned long long int> counter;
```
Comment (ezyang) [inline torch/csrc/generic/StorageSharing.cpp:45]: or just `uint64_t` hahaha

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
It seems this has broken some tests:

```
Dashboard
Projects
Insights
Organization Settings
Plan
Status
OPERATIONAL

Help
Edward Z. Yang

Dashboard
All Pipelines
Project
pytorch
Branch
pull/60978
Workflow
build
Job
pytorch_macos_10_13_py3_test (14487366)
pytorch_macos_10_13_py3_test
Failed

Rerun

Duration / Finished
16m 4s / 2d ago
Queued
9m 53s
Executor
macOS Medium

Branch
pull/60978
PR / Commit
#60978
/
a004e77
Author & Message
 
for shared-memory handles, use an atomic counter, instead of potentially colliding random numbers
22 tests failed
  out of 164
test_batch_sampler
test_multiple_dataloaders
test_partial_workers
test_sampler
test_seqential_batch_workers
test_seqential_batch_workers_prefetch
test_sequential_workers
test_shuffle_batch_workers
test_shuffle_batch_workers_prefetch
test_shuffle_reproducibility
test_shuffle_workers
test_batch_sampler
test_multiple_dataloaders
test_partial_workers
test_sampler
See all 22 in the tests tab below
Want to optimize your pipelines?
Find actionable data in
Test Insights
NEW
STEPS
TESTS
22
ARTIFACTS
22 tests failed
  out of 164 in python-unittest
test_batch_sampler - TestDataLoader
test_dataloader.py
Traceback (most recent call last):
  File "test_dataloader.py", line 1534, in test_batch_sampler
    self._test_batch_sampler(num_workers=4)
  File "test_dataloader.py", line 1522, in _test_batch_sampler
    for i, (input, _target) in enumerate(dl):
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 519, in __next__
    data = self._next_data()
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1201, in _next_data
    return self._process_data(data)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/dataloader.py", line 1227, in _process_data
    data.reraise()
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/_utils.py", line 425, in reraise
    raise self.exc_type(msg)
RuntimeError: Caught RuntimeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 52, in fetch
    return self.collate_fn(data)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 84, in default_collate
    return [default_collate(samples) for samples in transposed]
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 84, in <listcomp>
    return [default_collate(samples) for samples in transposed]
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 54, in default_collate
    storage = elem.storage()._new_shared(numel)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/storage.py", line 155, in _new_shared
    return cls._new_using_filename(size)
RuntimeError: falseINTERNAL ASSERT FAILED at "../aten/src/ATen/MapAllocator.cpp":243, please report a bug to PyTorch. unable to open shared memory object </torch_2722_0> in read-write mode
```

**Comment by mbc** [comment]
> It seems this has broken some tests:

Thanks, I will take a look.

**Comment by ezyang** [comment]
needs some modest portability fixes

```
Summary: 
xplat/caffe2/aten/src/ATen/MapAllocator.cpp:41:28: error: no viable conversion from 'pid_t' (aka 'int') to 'c10::guts::detail::DummyClassForToString'
  handle += std::to_string(getpid());
                           ^~~~~~~~
xplat/caffe2/aten/src/ATen/MapAllocator.cpp:44:28: error: no viable conversion from 'std::__atomic_base<unsigned long long>::__int_type' (aka 'unsigned long long') to 'c10::guts::detail::DummyClassForToString'
  handle += std::to_string(counter.fetch_add(1, std::memory_order_relaxed));
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by mbc** [comment]
@ezyang Thanks, checking. Is this generated from an FB-internal test? It seems gcc 4.8-related.

**Comment by ezyang** [comment]
yeah, it's some obscure internal build configuration for android. I'm not exactly sure how to get an exact compiler version

**Comment by mbc** [comment]
Thanks. It should - hopefully - be working now.

**Comment by mbc** [comment]
@ezyang Did it go OK internally this time?

**Comment by ezyang** [comment]
My bad I missed your update. Testing!

**Pull request closed without merging**